KEP-3243: Graduate MatchLabelKeys In PodTopologySpread to beta (#3740)
kubernetes/enhancements: keps/sig-scheduling/3243-respect-pod-topology-spread-after-rolling-upgrades/README.md

You can take a look at one potential example of such test in:
https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
-->
Yes. there are unit and integration tests for feature enablement/disablement.
- unit tests:
  - `pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go`
  - `pkg/scheduler/framework/plugins/podtopologyspread/scoring_test.go`
- integration tests
  -  `test/integration/scheduler/filters/filters_test.go`

>   -  `test/integration/scheduler/scoring/priorities_test.go`

Reviewer: These are all good tests that are testing the feature, but they don't test "enablement/disablement".

I would like to see a test that is exercising the enablement/disablement the feature - see the comment above for an example of such test:
https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282

[basically a unit test that changes the value of the feature gate in the middle of the test]

Author: Updated

rollout. Similarly, consider large clusters and how enablement/disablement
will rollout across nodes.
-->

> It won't impact already running workloads and it's an opt-in feature. 

Reviewer 2: ```suggestion
It won't impact already running workloads because it is an opt-in feature. 
```

> 

Reviewer 2: at line 643, do we still not have those tests?

Author: Updated.
We already have the unit and integration tests.

###### What specific metrics should inform a rollback?

<!--
What signals should users be paying attention to when the feature is young
that might indicate a serious problem?
-->
- If the metric `schedule_attempts_total{result="error|unschedulable"}` increased significantly after pods using this feature are added.

> - If the metric `plugin_execution_duration_seconds{plugin="PodTopologySpread"}` increased to larger than 100ms on 90% after pods using this feature are added. 

Reviewer 2: ```suggestion
- If the metric `plugin_execution_duration_seconds{plugin="PodTopologySpread"}` increased to higher than 100ms on 90% after pods using this feature are added. 
```

rollout. Similarly, consider large clusters and how enablement/disablement
will rollout across nodes.
-->

> It won't impact already running workloads because it is an opt-in feature.

Reviewer: It won't impact already running workloads because it's a scheduler feature - if something is already running, scheduler won't impact it for sure.

What about potential rollout issues?

Author: Updated

Longer term, we may want to require automated upgrade/rollback tests, but we
are missing a bunch of machinery and tooling and can't do that now.
-->

> Yes, it was tested manually following below steps, and get the expected result.

Reviewer 2: ```suggestion
Yes, it was tested manually by following the steps below, and it was working at intended.
```

1. create a kubernetes cluster v1.26 with 3 nodes where `MatchLabelKeysInPodTopologySpread` feature is disabled.
2. deploy a deployment with this yaml
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
spec:
  replicas: 12 
  selector:
    matchLabels:
      foo: bar
  template:
    metadata:
      labels:
        foo: bar
    spec:
      restartPolicy: Always
      containers:
      - name: nginx
        image: nginx:1.14.2
      topologySpreadConstraints:
        - maxSkew: 1
          topologyKey: kubernetes.io/hostname
          whenUnsatisfiable: DoNotSchedule
          labelSelector:
            matchLabels:
              foo: bar
          matchLabelKeys:
            - pod-template-hash
```
3. pods spread across nodes as 4/4/4
4. update the deployment nginx image to `nginx:1.15.0`
5. pods spread across nodes as 5/4/3
6. delete deployment nginx

> 7. upgrade kubenetes cluster to v1.27(with master branch) where `MatchLabelKeysInPodTopologySpread` feature is enable.

Reviewer 2: ```suggestion
7. upgrade kubenetes cluster to v1.27 (at master branch) while `MatchLabelKeysInPodTopologySpread` is enable.
```

8. deploy a deployment nginx like step2
9. pods spread across nodes as 4/4/4
10. update the deployment nginx image to `nginx:1.15.0`
11. pods spread across nodes as 4/4/4
12. delete deployment nginx

> 13. downgrade kubenetes cluster to v1.26  where `MatchLabelKeysInPodTopologySpread` feature is enable.

Reviewer 2: enabled?

checking if there are objects with field X set) may be a last resort. Avoid
logs or events for this purpose.
-->

> Operator can query pods that have the `pod.spec.topologySpreadConstraints.matchLabelKeys` field set to determin if the feature is in use by workloads. 

Reviewer 2: ```suggestion
Operator can query pods that have the `pod.spec.topologySpreadConstraints.matchLabelKeys` field set to determine if the feature is in use by workloads. 
```

Recall that end users cannot usually observe component logs or access metrics.
-->

- [x] Other (treat as last resort)

>   - Details: We can determine if the feature is being used by comparing the expected and actual scheduling results.

Reviewer: That's not particularly useful answer. How can I know "the expected scheduling result"?

Author: Updated

These goals will help you determine what you need to measure (SLIs) in the next
question.
-->

> Metric plugin_execution_duration_seconds{plugin="PodTopologySpread"} <= 100ms on 90-percentile.

Reviewer 2: are those numbers based on estimates mentioned in previous KEPs?

Author: Yes. It keeps the same numbers with previous KEPs.

[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
-->

> No.

Reviewer: The impact should be fairly small, but it actually impact scheduling latency, which directly contributes to pod-startup-latency SLO.

Author: Updated
